src/data/option/n_ary.lean

```diff
@@ -0,0 +1,153 @@
+
+@[simp] lemma map₂_some_some (f : α → β → γ) (a : α) (b : β) : map₂ f (some a) (some b) = f a b :=
+rfl
+lemma map₂_coe_coe (f : α → β → γ) (a : α) (b : β) : map₂ f a b = f a b := rfl
+@[simp] lemma map₂_none_left (f : α → β → γ) (b : option β) : map₂ f none b = none := rfl
+@[simp] lemma map₂_none_right (f : α → β → γ) (a : option α) : map₂ f a none = none :=
+by cases a; refl
+@[simp] lemma map₂_coe_left (f : α → β → γ) (a : α) (b : option β) :
+  map₂ f a b = b.map (λ b, f a b) := rfl
+@[simp] lemma map₂_coe_right (f : α → β → γ) (a : option α) (b : β) :
+  map₂ f a b = a.map (λ a, f a b) := rfl
+
+@[simp] lemma mem_map₂_iff {c : γ} : c ∈ map₂ f a b ↔ ∃ a' b', a' ∈ a ∧ b' ∈ b ∧ f a' b' = c :=
+by simp [map₂]
+
+@[simp] lemma map₂_eq_none_iff : map₂ f a b = none ↔ a = none ∨ b = none :=
+by cases a; cases b; simp
+
+lemma map₂_swap (f : α → β → γ) (a : option α) (b : option β) :
+  map₂ f a b = map₂ (λ a b, f b a) b a :=
+by cases a; cases b; refl
+
+lemma map_map₂ (f : α → β → γ) (g : γ → δ) : (map₂ f a b).map g = map₂ (λ a b, g (f a b)) a b :=
+by cases a; cases b; refl
+
+lemma map₂_map_left (f : γ → β → δ) (g : α → γ) :
+  map₂ f (a.map g) b = map₂ (λ a b, f (g a) b) a b :=
+by cases a; refl
+
+lemma map₂_map_right (f : α → γ → δ) (g : β → γ) :
+  map₂ f a (b.map g) = map₂ (λ a b, f a (g b)) a b :=
+by cases b; refl
+
+/-!
+### Algebraic replacement rules
+
+A collection of lemmas to transfer associativity, commutativity, distributivity, ... of operations
+to the associativity, commutativity, distributivity, ... of `option.map₂` of those operations.
+The proof pattern is `map₂_lemma operation_lemma`. For example, `map₂_comm mul_comm` proves that
+`map₂ (*) a b = map₂ (*) g f` in a `comm_semigroup`.
+-/
+
+lemma map₂_assoc {f : δ → γ → ε} {g : α → β → δ} {f' : α → ε' → ε} {g' : β → γ → ε'}
+  (h_assoc : ∀ a b c, f (g a b) c = f' a (g' b c)) :
+  map₂ f (map₂ g a b) c = map₂ f' a (map₂ g' b c) :=
+by cases a; cases b; cases c; simp [h_assoc]
+
+lemma map₂_comm {g : β → α → γ} (h_comm : ∀ a b, f a b = g b a) : map₂ f a b = map₂ g b a :=
+by cases a; cases b; simp [h_comm]
+
+lemma map₂_left_comm {f : α → δ → ε} {g : β → γ → δ} {f' : α → γ → δ'} {g' : β → δ' → ε}
+  (h_left_comm : ∀ a b c, f a (g b c) = g' b (f' a c)) :
+  map₂ f a (map₂ g b c) = map₂ g' b (map₂ f' a c) :=
+by cases a; cases b; cases c; simp [h_left_comm]
+
+lemma map₂_right_comm {f : δ → γ → ε} {g : α → β → δ} {f' : α → γ → δ'} {g' : δ' → β → ε}
+  (h_right_comm : ∀ a b c, f (g a b) c = g' (f' a c) b) :
+  map₂ f (map₂ g a b) c = map₂ g' (map₂ f' a c) b :=
+by cases a; cases b; cases c; simp [h_right_comm]
+
+lemma map_map₂_distrib {g : γ → δ} {f' : α' → β' → δ} {g₁ : α → α'} {g₂ : β → β'}
+  (h_distrib : ∀ a b, g (f a b) = f' (g₁ a) (g₂ b)) :
+  (map₂ f a b).map g = map₂ f' (a.map g₁) (b.map g₂) :=
+by cases a; cases b; simp [h_distrib]
+
+/-- Symmetric of `option.map₂_map_left_comm`. -/
```

Comment:
"Symmetric of" doesn't really make sense in English - I'm not sure what you mean here?

Reply:
Same equality but written the other way around.

If you wonder why we need this, it's because unification has a hard time figuring all the functions if you symmetrize on the spot. This is also how the other n-ary APIs do it.

Reply:
In that case, how about:
```suggestion
/-- Symmetric form of `option.map₂_map_left_comm`. -/
```
or
```suggestion
/-- Symmetric restatement of `option.map₂_map_left_comm`. -/
```
(also below)